File: src/lib/block.cpp
#include "bve_pch.h"
#include "block.h"
#include "lighting/spotlight.h"
#include "asset_manager.h"
namespace bve {
    block::~block() { }
    void block::load(ref<graphics::object_factory> object_factory, const namespaced_name& register_name) { }
    float block::opacity() { return 1.f; }
    bool block::solid() { return true; }
    ref<lighting::light> block::get_light() { return nullptr; }
    ref<model> block::get_model() { return nullptr; }
    bool block::managed() { return false; }
}

File: BasicVoxelEngine/Block.cs
using BasicVoxelEngine.Graphics;
using BasicVoxelEngine.Lighting;
using System;
using System.Runtime.CompilerServices;

namespace BasicVoxelEngine
{
    public abstract class Block : RegisteredObject<Block>
    {
        public Block() { }
        /// <summary>
        /// The opacity of this block. Default is 100% (1.0).
        /// </summary>
        public virtual float Opacity => 1f;
        /// <summary>
        /// If this block is solid or not. Default is true.
        /// </summary>
        public virtual bool Solid => true;
        /// <summary>
        /// The model that represents this block. Default is null.
        /// Use <see cref="Model.LoadModel(string, Factory)" /> to load a model.
        /// </summary>
        public virtual Model? Model => null;
        public virtual Light? Light => null;
        /// <summary>
        /// The friendly name of this block.
        /// </summary>
        public abstract string FriendlyName { get; }
    }
}